integration_tests/conftest.py

```diff
@@ -74,3 +80,189 @@ def pipeline_columns(sim, populations) -> List[str]:
+
+            no_bug_distribution = scipy.stats.betabinom(a=a, b=b, n=denominator)
+
+        bayes_factor = self._calculate_bayes_factor(
+            numerator, bug_distribution, no_bug_distribution
+        )
+
+        # TODO: Make this configurable?
+        reject_null = bayes_factor > 100
+        self.comparisons_made.append(
+            {
+                "name": name,
+                "name_addl": name_addl,
+                "proportion": proportion,
+                "numerator": numerator,
+                "denominator": denominator,
+                "target_value_lb": target_value_lb,
+                "target_value_ub": target_value_ub,
+                "bayes_factor": bayes_factor,
+                "reject_null": reject_null,
+            }
+        )
+
+        if reject_null:
+            if boolean_values.mean() < target_value_lb:
+                raise AssertionError(
+                    f"{name} value {proportion:g} is significantly less than expected, bayes factor = {bayes_factor:g}"
+                )
+            else:
+                raise AssertionError(
+                    f"{name} value {proportion:g} is significantly greater than expected, bayes factor = {bayes_factor:g}"
+                )
+
+        if (
+            target_value_lb > 0
+            and self._calculate_bayes_factor(0, bug_distribution, no_bug_distribution) < 100
+        ):
+            warnings.warn(
+                f"Sample size too small to ever find that the simulation's '{name}' value is less than expected."
+            )
+
+        if target_value_ub < 1 and (
+            self._calculate_bayes_factor(denominator, bug_distribution, no_bug_distribution)
+            < 100
+        ):
+            warnings.warn(
+                f"Sample size too small to ever find that the simulation's '{name}' value is greater than expected."
+            )
+
+        if 100 > bayes_factor > 0.1:
+            warnings.warn(f"Bayes factor for '{name}' is not conclusive.")
+
+    def _calculate_bayes_factor(
+        self,
+        numerator: int,
+        bug_distribution: scipy.stats.rv_discrete,
+        no_bug_distribution: scipy.stats.rv_discrete,
+    ) -> float:
+        # We can be dealing with some _extremely_ unlikely events here, so we have to set numpy to not error
+        # if we generate a probability too small to be stored in a floating point number(!), which is known
+        # as "underflow"
+        with np.errstate(under="ignore"):
+            bug_marginal_likelihood = bug_distribution.pmf(numerator)
+            no_bug_marginal_likelihood = no_bug_distribution.pmf(numerator)
+
+        try:
+            return bug_marginal_likelihood / no_bug_marginal_likelihood
+        except (ZeroDivisionError, FloatingPointError):
+            return 1_000_000.0
+
+    @cache
+    def _fit_beta_distribution_to_ui(self, lb: float, ub: float) -> Tuple[float, float]:
```

Comment:
How (or why) is this being cached? Doesn't this method have to run if you hit the branch in fuzzy_assert_proportions?

Reply:
This is being cached in memory, so if it is called multiple times in the same run with the same lb and ub parameters, it doesn't re-do the calculation, just uses the value it got last time.

In cases where we check something every timestep, for instance, this would prevent us from having to keep recalculating the beta for the same target UI over and over.

I didn't look at how much this cache is actually being used but it seems very low cost.